src/encord_active/lib/common/project.py

```diff
@@ -0,0 +1,245 @@
+from __future__ import annotations
+
+import itertools
+import json
+import logging
+from functools import partial
+from pathlib import Path
+from typing import Dict, List, Optional, Union
+
+import encord.exceptions
+import yaml
+from encord import Project as EncordProject
+from encord.objects.ontology_structure import OntologyStructure
+from encord.orm.label_row import LabelRow
+from encord.project import LabelRowMetadata
+
+from encord_active.lib.common.utils import (
+    collect_async,
+    download_file,
+    fetch_project_meta,
+    slice_video_into_frames,
+)
+
+logger = logging.getLogger(__name__)
+encord_logger = logging.getLogger("encord")
+encord_logger.setLevel(logging.ERROR)
+
+
+class Project:
+    def __init__(self, project_dir: Path, subset_size: Optional[int] = None):
+        """
+        Fetch project's data from local storage.
+
+        :param project_dir: The root directory of the project.
+        :param subset_size: The number of label rows to fetch. If None then all label rows are fetched.
+        """
+        if not project_dir.exists():
+            raise FileNotFoundError("`project_dir` must point to an existing directory")
+        if not project_dir.is_dir():
+            raise NotADirectoryError("`project_dir` must point to an existing directory")
+        self.project_dir: Path = project_dir
+
+        # read project's metadata and hash
+        self.project_meta: Dict[str, str] = fetch_project_meta(project_dir)
+        self.project_hash: str = self.project_meta["project_hash"]
+
+        # read project ontology
+        ontology_file_path = project_dir / "ontology.json"
+        if not ontology_file_path.exists():
+            raise FileNotFoundError(f"Expected file `ontology.json` at {project_dir}")
+        self.ontology: OntologyStructure = OntologyStructure.from_dict(
+            json.loads((project_dir / "ontology.json").read_text())
+        )
+
+        # read label rows' metadata
+        label_row_meta_file_path = project_dir / "label_row_meta.json"
+        if not label_row_meta_file_path.exists():
+            raise FileNotFoundError(f"Expected file `label_row_meta.json` at {project_dir}")
+        self.label_row_meta: Dict[str, LabelRowMetadata] = {}
+        for lr_hash, lr_meta in itertools.islice(json.loads(label_row_meta_file_path.read_text()).items(), subset_size):
+            self.label_row_meta[lr_hash] = LabelRowMetadata.from_dict(lr_meta)
+
+        # read label rows and their images
+        self.label_rows: Dict[str, LabelRow] = {}
+        self.image_paths: Dict[str, List[Path]] = {}
```

Comment:
The way it works right now is that the iterator, for example, will take the stem of that path as the data unit hash so it will do the comparison there. We probably should take that off the iterator and convert `List[Path]` to a dict of data unit hash to image path.

Reply:
Yes that sound much better 👍